### PR TITLE
Flash editable block outlines instead of always showing them

### PR DESCRIPTION
**Pull request by noisysocks**
## What?
Iterates on https://github.com/WordPress/gutenberg/pull/57901.

Updates the editable block outlines that appear within content-locked containers (added in #57901) to appear and then fade out after 1s when:

- The page loads; or
- The user clicks on the content-locked container.

## Why?
See discussion in https://github.com/WordPress/gutenberg/pull/57901#issuecomment-1904339567. We want to reduce the prominence of these outlines as they are currently too distracting from the writing experience. 

## How?
A `.has-editable-block-outline` class is added to blocks that are editable but within a locked container. This class shows the outline and then fades it away after 1s.

Then, a private `useFlashEditableBlocks()` hook is attached to relevant containers: template-locked pages, synced patterns, and patterns with `templateLock = 'contentOnly'`. The hook listens for a `click` on a disabled element and replays the animation when this happens.

## Testing Instructions
The outlines should should appear when editing a content only locked pattern, a pattern with instance overrides, and a page in the site editor.

Content only locked patterns:

1. Add a pattern to your current theme that has a `contentOnly` lock. [Here's an example of one.](https://gist.github.com/richtabor/ddeea41ced691721318649bea8ce9db8)
2. Edit a page or template and insert the pattern.

Pattern with instance overrides:

1. Follow [these instructions](https://github.com/WordPress/gutenberg/issues/53705#issuecomment-1882305331) to create a pattern with instance overrides enabled.
2. Edit a page or template and insert the pattern.

Page in the site editor:

1. Go to Appearance → Editor → Pages and select a page to edit.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/612155/4efbaff2-bd0c-4f73-808e-55a200daedb0

### CODE REVIEW

**Comment by SaxonF** [comment]
This is an improvement over what's in trunk so LGTM once we're happy with the technical side of things. We can keep iterating until we find the right balance between visibility and it being a distraction.

**Comment by jameskoster** [comment]
Isn't it _supposed_ to be a distraction? Almost feels like the treatment could be heavier / more obvious.

One observation (probably doesn't need to hold up this effort): the effects when clicking non-editable content seem contradictory. The Snackbar suggests editing the template, while the outlines imply editing the content. This feels a bit unsettling, like being pulled in opposite directions.

**Comment by jasmussen** [comment]
This is an improvement, thanks for sweating the details. That said, it wasn't entirely what I expected, I think we can finetune it a little bit. Specifically, I wouldn't expect it to be present _from the start_ — that especially makes those borders appear like content, rather than editor guides. Secondly, the fading and time before fading out feels very slow, I think that actually detracts from the emphasis they are meant to provide.

The behavior I was expecting was that you click to edit a page, and if you never hover anything other than the Post Content block, you won't see them. But as soon as you hover the header, or the footer, or something other than the post content, those outlines fade in _fast_, and very shortly after, fade out. They might even pulse, if need be, to provide all the more emphasis. In them appearing fast, pulsing, then disappearing, and only when hovering other-than-post-content, there would be an immediate connection to the behavior: _no, not here, down there is where you edit_. Would love @richtabor thoughts too.

Separately, I still find it a bit confusing that the post content block is full-width, when the title block is not. Is this a result of the width of the elements in the DOM, or an intentional choice?

**Comment by jameskoster** [comment]
> The behavior I was expecting was that you click to edit a page, and if you never hover anything other than the Post Content block, you won't see them. 

> But as soon as you hover the header, or the footer, or something other than the post content

In practise I fear this would mean you'd be seeing the outlines much more often. For empty pages in particular, the majority of the canvas will be 'template'.

> Separately, I still find it a bit confusing that the post content block is full-width, when the title block is not. Is this a result of the width of the elements in the DOM, or an intentional choice?

This will be due to the template block structure. E.g. Title may be in a Group where Content is root level.

**Comment by noisysocks** [comment]
> Separately, I still find it a bit confusing that the post content block is full-width, when the title block is not. Is this a result of the width of the elements in the DOM, or an intentional choice?

We're highlighting the border box (what's clickable). We could probably highlight the content box (what an inner block could expand to fill) with some work—we'd have to stop using pseudo `:after` elements. But doesn't it make sense that the outline appears around what the user can click?

**Comment by noisysocks** [comment]
I finished the remaining TODOs, rebased this, and reduced the animation delay to 1s.

**Comment by noisysocks** [comment]
I think let's get this in as it's a big improvement on what's already in `trunk`, and then we can continue iterating on further ideas. (Like [this good one](https://github.com/WordPress/gutenberg/pull/58232#issuecomment-1918039323).) Just waiting for a code review now.

File: packages/block-editor/src/components/use-flash-editable-blocks/index.js
Comment (andrewserong) [inline packages/block-editor/src/components/use-flash-editable-blocks/index.js:38]: Oh, clever!

File: packages/editor/src/components/editor-canvas/edit-template-blocks-notification.js
Comment (andrewserong) [inline packages/editor/src/components/editor-canvas/edit-template-blocks-notification.js:63]: Just double-checking that I'm following correctly: the `renderingMode` is removed from the checks in the click and double click handler because we're moving the check up to where the component is being output instead of making it internal to the component?

**Review by andrewserong — APPROVED** [review]
Code looks great, and behaviour-wise this feels terrific to me! 1 second fade feels like a nice amount of time, and it feels like a good level of responsiveness while being unobtrusive to me when I go to click on the template.

✅ Works with partially synced patterns
✅ Works with content only locked patterns
✅ Works with pages edited in the site editor
✅ Works with posts in the post editor in template preview mode

https://github.com/WordPress/gutenberg/assets/14988353/4a9687f7-77ed-448e-b80d-309f7a5eb1bf

LGTM! ✨

**Comment by jameskoster** [comment]
The outline behavior is working pretty well for me. But I still find them _plus_ the Snackbar too much. The outlines disappear before I have time to read the Snackbar message, so each affordance effectively detracts from the other. Is it worth trying a version without the Snackbar?

File: packages/editor/src/components/editor-canvas/edit-template-blocks-notification.js
Comment (noisysocks) [reply packages/editor/src/components/editor-canvas/edit-template-blocks-notification.js:63]: Yeah you have it right.

It's not necessary for this PR at all—leftover code from some explorations I was doing that went nowhere.

But I kept these changes as I think it's much nicer to simply not mount the component at all if it's not needed rather than unnecessarily mount the component and attach all these unnecessary event listeners that run unnecessary code on every click.

**Comment by noisysocks** [comment]
> The outline behavior is working pretty well for me. But I still find them _plus_ the Snackbar too much. The outlines disappear before I have time to read the Snackbar message, so each affordance effectively detracts from the other. Is it worth trying a version without the Snackbar?

Yeah I'm not fond of the snackbar notifications now that we have the outlines. The only thing that gives me pause re. removing them is that it's currently the most convenient way of accessing the template. I think we'd need to make the _Edit template_ button (currently buried underneath the Template menu in the document settings) more prominent.

Was it you @jameskoster that had a mockup of putting that button in the View menu alongside Desktop, Tablet, Mobile? I like that. What do you think @SaxonF?

At any rate, let's keep firing away on all that in follow-up PRs 😀

**Comment by jameskoster** [comment]
Yeah there are so many options to explore there, let's continue refining in subsequent issues/prs.

**Pull request merged**
